File: LoanCalculator/Models/LoanCalculator.swift
import Foundation

struct LoanResult: Equatable {
    let periodicPayment: Double
    let totalInterest: Double
    let totalPaid: Double
    let schedule: [AmortizationPayment]
}

struct AmortizationPayment: Identifiable, Equatable {
    let id: Int
    let paymentNumber: Int
    let paymentAmount: Double
    let interestPaid: Double
    let principalPaid: Double
    let remainingBalance: Double
}

enum LoanCalculator {
    static func calculate(for parameters: LoanParameters) -> LoanResult {
        let principal = max(parameters.loanAmount, 0)
        let paymentsPerYear = Double(parameters.paymentsPerYear.rawValue)
        let numberOfPayments = Int((parameters.termYears * paymentsPerYear).rounded())

        guard numberOfPayments > 0 else {
            return LoanResult(
                periodicPayment: 0,
                totalInterest: 0,
                totalPaid: 0,
                schedule: []
            )
        }

        let ratePerPeriod = (parameters.interestRate / 100) / paymentsPerYear
        let paymentAmount: Double

        if ratePerPeriod.isZero {
            paymentAmount = principal / Double(numberOfPayments)
        } else {
            let factor = pow(1 + ratePerPeriod, Double(numberOfPayments))
            paymentAmount = principal * (ratePerPeriod * factor) / (factor - 1)
        }

        var remainingBalance = principal
        var schedule: [AmortizationPayment] = []
        var totalInterest: Double = 0
        var totalPaid: Double = 0

        for paymentNumber in 1...numberOfPayments {
            let interestPayment = remainingBalance * ratePerPeriod
            var principalPayment = paymentAmount - interestPayment
            principalPayment = min(max(principalPayment, 0), remainingBalance)

            if paymentNumber == numberOfPayments {
                principalPayment = remainingBalance
            }

            let actualPayment = principalPayment + max(interestPayment, 0)

            remainingBalance -= principalPayment
            remainingBalance = max(remainingBalance, 0)

            totalInterest += max(interestPayment, 0)
            totalPaid += actualPayment

            let amortizationPayment = AmortizationPayment(
                id: paymentNumber,
                paymentNumber: paymentNumber,
                paymentAmount: actualPayment,
                interestPaid: max(interestPayment, 0),
                principalPaid: principalPayment,
                remainingBalance: remainingBalance
            )
            schedule.append(amortizationPayment)
        }

        return LoanResult(
            periodicPayment: paymentAmount,
            totalInterest: totalInterest,
            totalPaid: totalPaid,
            schedule: schedule
        )
    }
}

File: src/models/LoanCalculator.ts
import { LoanParameters } from './LoanParameters';

export interface AmortizationPayment {
  id: number;
  paymentNumber: number;
  paymentAmount: number;
  interestPaid: number;
  principalPaid: number;
  remainingBalance: number;
}

export interface LoanResult {
  periodicPayment: number;
  totalInterest: number;
  totalPaid: number;
  schedule: AmortizationPayment[];
}

export class LoanCalculator {
  static calculate(parameters: LoanParameters): LoanResult {
    const principal = Math.max(parameters.loanAmount, 0);
    const paymentsPerYear = parameters.paymentsPerYear;
    const numberOfPayments = Math.round(parameters.termYears * paymentsPerYear);

    if (numberOfPayments <= 0) {
      return {
        periodicPayment: 0,
        totalInterest: 0,
        totalPaid: 0,
        schedule: [],
      };
    }

    const ratePerPeriod = (parameters.interestRate / 100) / paymentsPerYear;
    let paymentAmount: number;

    if (ratePerPeriod === 0) {
      paymentAmount = principal / numberOfPayments;
    } else {
      const factor = Math.pow(1 + ratePerPeriod, numberOfPayments);
      paymentAmount = principal * (ratePerPeriod * factor) / (factor - 1);
    }

    let remainingBalance = principal;
    const schedule: AmortizationPayment[] = [];
    let totalInterest = 0;
    let totalPaid = 0;

    for (let paymentNumber = 1; paymentNumber <= numberOfPayments; paymentNumber++) {
      const interestPayment = remainingBalance * ratePerPeriod;
      let principalPayment = paymentAmount - interestPayment;
      principalPayment = Math.min(Math.max(principalPayment, 0), remainingBalance);

      if (paymentNumber === numberOfPayments) {
        principalPayment = remainingBalance;
      }

      const actualPayment = principalPayment + Math.max(interestPayment, 0);

      remainingBalance -= principalPayment;
      remainingBalance = Math.max(remainingBalance, 0);

      totalInterest += Math.max(interestPayment, 0);
      totalPaid += actualPayment;

      const amortizationPayment: AmortizationPayment = {
        id: paymentNumber,
        paymentNumber,
        paymentAmount: actualPayment,
        interestPaid: Math.max(interestPayment, 0),
        principalPaid: principalPayment,
        remainingBalance,
      };
      schedule.push(amortizationPayment);
    }

    return {
      periodicPayment: paymentAmount,
      totalInterest,
      totalPaid,
      schedule,
    };
  }
}